### PR TITLE
Make ProxyTypes work on Mono

### PR DESCRIPTION
**Pull request by ryanbrandenburg**

### CODE REVIEW

File: src/Microsoft.Extensions.DiagnosticAdapter/Internal/ProxyTypeEmitter.cs [hunk 1]
```diff
@@ -260,6 +260,9 @@ private static bool VerifyProxySupport(ProxyBuilderContext context, Tuple<Type,
             il.Emit(OpCodes.Ldarg_0);
             il.Emit(OpCodes.Ldarg_1);
             il.Emit(OpCodes.Ldtoken, proxyType);
+
+            var getTypeFromHandle = typeof(Type).GetMethod("GetTypeFromHandle", BindingFlags.Static | BindingFlags.Public);
+            il.EmitCall(OpCodes.Call, getTypeFromHandle, null);
```
Comment (rynowak) [inline src/Microsoft.Extensions.DiagnosticAdapter/Internal/ProxyTypeEmitter.cs]: Is this what the C# compiler does when it needs to do a `typeof(...)`?

File: src/Microsoft.Extensions.DiagnosticAdapter/Internal/ProxyTypeEmitter.cs [hunk 2]
```diff
@@ -322,38 +325,54 @@ private static void GenerateProxyTypeFromProperties(Type sourceType, Type target
                 {
                     il.DeclareLocal(targetProperty.PropertyType);
 
-                    // Return a default(T) value.
                     il.Emit(OpCodes.Ldloca_S, 0);
                     il.Emit(OpCodes.Initobj, targetProperty.PropertyType);
 
+                    // Return a default(T) value.
```
Comment (rynowak) [inline src/Microsoft.Extensions.DiagnosticAdapter/Internal/ProxyTypeEmitter.cs]: This should go back to the top of the block

File: src/Microsoft.Extensions.DiagnosticAdapter/Internal/ProxyTypeEmitter.cs [hunk 3]
```diff
@@ -322,38 +325,54 @@ private static void GenerateProxyTypeFromProperties(Type sourceType, Type target
...
                     // If it returned null, we want to just jump to return.
-                    il.Emit(OpCodes.Dup);
+                    il.Emit(OpCodes.Ldloc_S, 1);
                     il.Emit(OpCodes.Brfalse, @return);
+                }
 
-                    // Create a proxy for the value returned by source property (if necessary).
-                    EmitProxy(context, il, targetProperty.PropertyType, sourceProperty.PropertyType);
+                il.Emit(OpCodes.Ldloc_S, 1);
+                // Create a proxy for the value returned by source property (if necessary).
```
Comment (rynowak) [inline src/Microsoft.Extensions.DiagnosticAdapter/Internal/ProxyTypeEmitter.cs]: Move this up one line

File: src/Microsoft.Extensions.DiagnosticAdapter/Internal/ProxyTypeEmitter.cs [hunk 4]
```diff
@@ -322,38 +325,54 @@ private static void GenerateProxyTypeFromProperties(Type sourceType, Type target
...
+                        "Instance",
+                        BindingFlags.Instance | BindingFlags.Public));
 
+                // Call the source property.
+                il.EmitCall(OpCodes.Callvirt, sourceProperty.GetMethod, null);
+                il.Emit(OpCodes.Stloc_S, 1);
+
+                var @return = il.DefineLabel();
+
+                if (!sourceProperty.PropertyType.GetTypeInfo().IsValueType)
```
Comment (rynowak) [inline src/Microsoft.Extensions.DiagnosticAdapter/Internal/ProxyTypeEmitter.cs]: Needs a comment explaining why/what

File: src/Microsoft.Extensions.DiagnosticAdapter/Internal/ProxyAssembly.cs [hunk 5]
```diff
@@ -41,7 +41,7 @@ static ProxyAssembly()
             return ModuleBuilder.DefineType(name, attributes, baseType, interfaces);
         }
 
-#if GENERATE_ASSEMBLIES
+#if NET451 && GENERATE_ASSEMBLIES
```
Comment (rynowak) [inline src/Microsoft.Extensions.DiagnosticAdapter/Internal/ProxyAssembly.cs]: ???

Obvi this functionality isn't available on CoreCLR. I solved this in the past by just defining this for net451

File: src/Microsoft.Extensions.DiagnosticAdapter/Internal/ProxyTypeEmitter.cs (hunk 1 above)
Comment (ryanbrandenburg) [inline src/Microsoft.Extensions.DiagnosticAdapter/Internal/ProxyTypeEmitter.cs]: Yes. I had assumed there had to be some cleaner way to do it too, but that's what it put out.

Comment (rynowak) [inline src/Microsoft.Extensions.DiagnosticAdapter/Internal/ProxyTypeEmitter.cs]: Make explain with comments

**Comment by ryanbrandenburg** [comment]
:up: :date:

File: src/Microsoft.Extensions.DiagnosticAdapter/Internal/ProxyTypeEmitter.cs [hunk 6]
```diff
@@ -320,40 +326,58 @@ private static void GenerateProxyTypeFromProperties(Type sourceType, Type target
...
+                        BindingFlags.Instance | BindingFlags.Public));
+
+                // Call the source property.
+                il.EmitCall(OpCodes.Callvirt, sourceProperty.GetMethod, null);
+                il.Emit(OpCodes.Stloc_S, 1);
+
+                var @return = il.DefineLabel();
+
+                // BrFalse can't handle all value types (Decimal for example), so we skip it for value types
+                // The proxy will do nothing so it's fine
```
Comment (rynowak) [inline src/Microsoft.Extensions.DiagnosticAdapter/Internal/ProxyTypeEmitter.cs]: `The proxy will do nothing so it's fine` - actually the issue here is that we really are using `brfalse` as a null-check for reference types. If it's a value type we don't need to check for null :+1:

**Comment by rynowak** [comment]
:shipit: after fixing one test

**Comment by rynowak** [comment]
:shipit:

**Pull request merged**